### PR TITLE
[Enhacne] Support maybe_raise_or_warn for peft

### PR DESCRIPTION
**Pull request by okotaku**
# What does this PR do?

Issues #5714 

```
import torch
from diffusers import DiffusionPipeline, UNet2DConditionModel, AutoencoderKL
from peft import LoraConfig, get_peft_model

prompt = 'a man'

unet = UNet2DConditionModel.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", subfolder='unet', torch_dtype=torch.float16)
unet = get_peft_model(unet, LoraConfig(target_modules=["to_q", "to_v", "query", "value"]))
vae = AutoencoderKL.from_pretrained(
    'madebyollin/sdxl-vae-fp16-fix',
    torch_dtype=torch.float16,
)
pipe = DiffusionPipeline.from_pretrained(
    'stabilityai/stable-diffusion-xl-base-1.0', unet=unet, vae=vae, torch_dtype=torch.float16)
pipe.to('cuda')

image = pipe(
    prompt,
    num_inference_steps=50,
    width=1024,
    height=1024,
).images[0]
image.save('demo.png')
```

This code raised following error.

```
 ValueError: PeftModel() is of type: <class 'peft.peft_model.PeftModel'>, but should be <class 'diffusers.models.modeling_utils.ModelMixin'>
```

I solved it.

With this PR, we can fix the following lines to pass the "unet" directory to the pipeline.

https://github.com/huggingface/peft/blob/main/examples/lora_dreambooth/train_dreambooth.py#L992-L1000

```
pipeline = DiffusionPipeline.from_pretrained(
                        args.pretrained_model_name_or_path,
                        unet=unet,
                        text_encoder=text_encoder,
                        safety_checker=None,
                        revision=args.revision,
                    )
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by DN6** [comment]
cc: @sayakpaul @younesbelkada

**Comment by yiyixuxu** [comment]
@sayakpaul could you take a look here?

File: src/diffusers/pipelines/pipeline_utils.py [hunk 1]
```diff
@@ -288,9 +289,15 @@ def maybe_raise_or_warn(
...
         # I didn't find a public API to get the original class.
         sub_model = passed_class_obj[name]
-        model_cls = sub_model.__class__
         if is_compiled_module(sub_model):
-            model_cls = sub_model._orig_mod.__class__
+            sub_model = sub_model._orig_mod
+
+        model_cls = sub_model.__class__
+        if USE_PEFT_BACKEND:
+            from peft import PeftModel
```
Comment (younesbelkada) [inline src/diffusers/pipelines/pipeline_utils.py:297]: Not sure if this condition is needed, I would rather do something like:
```suggestion
        if is_peft_available() and isinstance(sub_model, PeftModel):
            from peft import PeftModel
```
As we don't use `PeftModel` for the PEFT integration, it should be totally agnostic to the minimum PEFT version we require for the integration.

**Review by younesbelkada — COMMENTED** [review]
In principle this looks good! I left one comment, IMO this PR should be agnostic to PEFT integration and fixes issues when users pass a `PeftModel`. Not sure also how much this is in the scope of diffusers as we might want users to use the PEFT integration through the low level API rather than using `PeftModel` interface

**Comment by younesbelkada** [comment]
You could also do something different by exposing a method `unwrap_model` that takes care of the whole logic if the model is a peft model, etc. and use that method that can be also copied over different classes / modules

**Comment by okotaku** [comment]
In the training code in the examples, unet is passed to the pipeline when the validation is run. 

https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora_sdxl.py#L1146

It would be nice to be able to pass the PeftModel directly to the pipeline at that time, eliminating the overhead for the user.

**Comment by sayakpaul** [comment]
We'll soon start working on refactoring our training scripts to support that :-) 

@younesbelkada any other comments on this PR?

**Comment by sayakpaul** [comment]
> With this PR, we can fix the following lines to pass the "unet" directory to the pipeline.

Hmm, I think we always want the users to load the `unet` first and then init a pipeline with it. I am not sure if we'd want to deviate from that design TBH. Hence, I'd like an opinion from @patrickvonplaten here.

**Review by younesbelkada — APPROVED** [review]
Makes sense thanks! I will let diffusers maintainers give their final approval
As it would be a quite common scenario to unwrap the model I would define an utility method `unwrap_model`

```python
def unwrap_model(model):
       if is_compiled_module(model):
            model = model._orig_mod

       if is_peft_available() and isinstance(model, PeftModel):
            model = model.base_model.model

       return model
```

**Comment by patrickvonplaten** [comment]
@okotaku could we maybe create a `unwrap_model` method as mentioned by @younesbelkada here: https://github.com/huggingface/diffusers/pull/5653#pullrequestreview-1724702190 We already have 3 occurances where we unwrapt the model and given that we now have a another model class wrapper it'd be good to add a utility function:

```
src/diffusers/pipelines/pipeline_utils.py
292:            model_cls = sub_model._orig_mod.__class__
550:                    not_compiled_module = module._orig_mod
655:                sub_model = sub_model._orig_mod
```

**Comment by okotaku** [comment]
I have fixed.

**Comment by patrickvonplaten** [comment]
Very nice!

**Comment by patrickvonplaten** [comment]
Follow-up PR: https://github.com/huggingface/diffusers/pull/5789

**Pull request merged**